### DISCUSSION TITLE
vipsthumbnail fails with piped stdin, but not with stdin as file descriptor

### DISCUSSION THREAD

**Discussion by martoche** [libvips/libvips#4304] [Q&A]
## Bug report

**Describe the bug**
The `vipsthumbnail` fails with a piped stdin, but not with stdin as a file descriptor.

In consequence, `vipsthumbnail` is not very friendly to use when being exec'd from another program that passes the data directly from memory. Eg: a webserver that has an image in memory and wants to make a thumbnail.

**To Reproduce**
Run the following command in a terminal:
```
$ cat test.heic | vipsthumbnail stdin > test.jpg

(vipsthumbnail:20719): VIPS-WARNING **: 17:20:55.771: error in tile 0 x 0
vipsthumbnail: unable to thumbnail stdin
source: bad seek to 461
heif: Invalid input: Unspecified: Error setting input file position (2.0)
```

**Expected behavior**
I expected the above command to work, just like this one works:
```
$ vipsthumbnail stdin < test.heic > test.jpg
```

**Actual behavior**
An error is printed in stderr, see above.

**Environment**
(please complete the following information)
 - OS: macOS 13.7.1
 - Vips: 8.16.0

**Comment by jcupitt** [comment]
Hi @martoche, 

This happens for formats which need to be able to seek their input, like HEIC and TIFF. You'll find thumbnail works for all JPG images (no seeking required for this format):

```
$ cat test.jpg | vipsthumbnail stdin -o .jpg > x.jpg
```

If you try this with something like HEIC, libvips will see the seek and automatically switch to buffering the input image in memory. However, this is a security risk, since an attacker could send unbounded data down the pipe and fill server ram.

With small images, this should just work:

```
$ cat image1.heic | vipsthumbnail stdin -o .jpg > x.jpg
```

With very large images (over 1gb I think) you will need to set `VIPS_PIPE_READ_LIMIT`, perhaps:

```
$ VIPS_PIPE_READ_LIMIT=10gb cat image1.heic | vipsthumbnail stdin -o .jpg > x.jpg
```

**Reply by jcupitt** [reply]
... though I wonder if you have hit that limit. Could you link a sample HEIC that fails? All my HEICs seem to work.

**Comment by martoche** [comment]
Hi @jcupitt, thanks a lot for your answer.

The problem happens with a 1 pixel image (504 bytes) that I use for testing that I've created with gimp. I've attached the file below (sorry, I had to put it in a zip file otherwise Github won't allow it):

[1px.zip](https://github.com/user-attachments/files/18012894/1px.zip)

**Reply by jcupitt** [reply]
Thanks! See the error here too on macOS, but not on linux where it seems to work fine. I suppose there's been some change in libheif that's caused this to show up.

I'll have a look.
